cmd/portforward.go

```diff
@@ -0,0 +1,548 @@
+	stop         chan struct{}
+	err          error
+}
+
+func getPortMappings(args []string) ([]portMapping, error) {
+	var err error
+	portMappings := []portMapping{}
+	for _, arg := range args {
+		remoteHost := "127.0.0.1"
+		protocol := protocolTCP
+		if strings.Contains(arg, "/") {
+			parts := strings.SplitN(arg, "/", 2)
+			if parts[0] == protocolTCP {
+				protocol = protocolTCP
+			} else if parts[0] == protocolUDP {
+				protocol = protocolUDP
+			} else {
+				return nil, errors.New("unknown protocol: " + parts[0])
+			}
+			arg = parts[1]
+		}
+		var localPort, remotePort int
+		if strings.Contains(arg, ":") {
+			parts := strings.SplitN(arg, ":", 3)
+			if len(parts) == 3 {
+				remoteHost = parts[0]
+				parts = parts[1:]
+			}
+			localPort, err = strconv.Atoi(parts[0])
+			if err != nil || localPort < 0 || localPort > 65536 {
+				return nil, errors.New("invalid port number: " + parts[0])
+			}
+			remotePort, err = strconv.Atoi(parts[1])
+			if err != nil || remotePort < 0 || remotePort > 65536 {
+				return nil, errors.New("invalid port number: " + parts[1])
+			}
+		} else {
+			port, err := strconv.Atoi(arg)
+			if err != nil || port < 0 || port > 65536 {
+				return nil, errors.New("invalid port number: " + arg)
+			}
+			localPort = port
+			remotePort = port
+		}
+		portMappings = append(portMappings, portMapping{
+			Protocol:   protocol,
+			LocalPort:  uint(localPort),
+			RemoteHost: remoteHost,
+			RemotePort: uint(remotePort),
+		})
+	}
+	return portMappings, nil
+}
+
+// NewPortForwardCmd returns a new PortForwardCmd
+func NewPortForwardCmd(cmd *cobra.Command, args []string) (*PortForwardCmd, error) {
+	server, err := cmd.Flags().GetString(argRootServer)
+	if err != nil {
+		return nil, err
+	}
+
+	skipVerify, err := cmd.Flags().GetBool(argRootSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	bindingHost, err := cmd.Flags().GetString(argBindHost)
+	if err != nil {
+		return nil, err
+	}
+
+	portMappings, err := getPortMappings(args[1:])
+	if err != nil {
+		return nil, err
+	}
+
+	return &PortForwardCmd{
+		server:       server,
+		skipVerify:   skipVerify,
+		deviceID:     args[0],
+		bindingHost:  bindingHost,
+		portMappings: portMappings,
+		connections:  make(map[string]chan *ws.ProtoMsg),
+		stop:         make(chan struct{}),
+	}, nil
+}
+
+func (c *PortForwardCmd) getToken() ([]byte, error) {
+	tokenPath, err := getDefaultAuthTokenPath()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to determine the auth token path")
+	}
+	token, err := ioutil.ReadFile(tokenPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "Please Login first")
+	}
+	return token, nil
+}
+
+type TCPPortForwarder struct {
+	listen     net.Listener
+	remoteHost string
+	remotePort uint
+}
+
+func NewTCPPortForwarder(bindingHost string, localPort uint, remoteHost string, remotePort uint) (*TCPPortForwarder, error) {
+	fmt.Printf("Forwarding from %s:%d -> %s:%d\n", bindingHost, localPort, remoteHost, remotePort)
```

Comment:
do we really have no logger in the cli? (sorry have not checked)

Reply:
no logger